README.md
# SCRU128: Sortable, Clock and Random number-based Unique identifier

[![GitHub tag](https://img.shields.io/github/v/tag/scru128/go-scru128)](https://github.com/scru128/go-scru128)
[![License](https://img.shields.io/github/license/scru128/go-scru128)](https://github.com/scru128/go-scru128/blob/main/LICENSE)

SCRU128 ID is yet another attempt to supersede [UUID] for the users who need
decentralized, globally unique time-ordered identifiers. SCRU128 is inspired by
[ULID] and [KSUID] and has the following features:

- 128-bit unsigned integer type
- Sortable by generation time (as integer and as text)
- 25-digit case-insensitive textual representation (Base36)
- 48-bit millisecond Unix timestamp that ensures useful life until year 10889
- Up to 281 trillion time-ordered but unpredictable unique IDs per millisecond
- 80-bit three-layer randomness for global uniqueness

```go
import "github.com/scru128/go-scru128/v3"

// generate a new identifier object
x := scru128.New()
fmt.Println(x)    // e.g., "036z951mhjikzik2gsl81gr7l"
fmt.Println(x[:]) // as a 128-bit unsigned integer in big-endian byte array

// generate a textual representation directly
fmt.Println(scru128.NewString()) // e.g., "036z951mhzx67t63mq9xe6q0j"
```

See [SCRU128 Specification] for details.

[UUID]: https://en.wikipedia.org/wiki/Universally_unique_identifier
[ULID]: https://github.com/ulid/spec
[KSUID]: https://github.com/segmentio/ksuid
[SCRU128 Specification]: https://github.com/scru128/spec

## License

Licensed under the Apache License, Version 2.0.

## See also

- [scru128 package - github.com/scru128/go-scru128/v3 - Go Packages](https://pkg.go.dev/github.com/scru128/go-scru128/v3)


CHANGELOG.md
# Changelog

## v3.0.2 - 2023-09-17

### Added

- Early `nil` checks to `Generator` methods

### Maintenance

- Improved documentation about generator's clock rollback behavior
- Renamed source files: generator -> gen, identifier -> id

## v3.0.1 - 2023-07-17

Most notably, v3 switches the letter case of generated IDs from uppercase (e.g.,
"036Z951MHJIKZIK2GSL81GR7L") to lowercase (e.g., "036z951mhjikzik2gsl81gr7l"),
though it is technically not supposed to break existing code because SCRU128 is
a case-insensitive scheme. Other changes include the removal of deprecated APIs.

### Removed

- Deprecated items:
  - `Generator#GenerateCore()`
  - `Generator#LastStatus()` and `GeneratorStatus`

### Changed

- Letter case of generated IDs from uppercase to lowercase
- Edge case behavior of generator functions' rollback allowance handling

### Maintenance

- Upgraded minimum Go version to 1.20

## v2.3.2 - 2023-06-21

### Changed

- Error values returned by `Generator` and `Id` to improve error messages

## v2.3.1 - 2023-04-07

### Maintenance

- Tweaked docs and tests

## v2.3.0 - 2023-03-23

### Added

- `GenerateOrAbort()` and `GenerateOrAbortCore()` to `Generator` (formerly named
  as `GenerateNoRewind()` and `GenerateCoreNoRewind()`)
- `Generator#GenerateOrResetCore()`

### Deprecated

- `Generator#GenerateCore()`
- `Generator#LastStatus()` and `GeneratorStatus`

## v2.2.1 - 2023-03-19

### Added

- `GenerateNoRewind()` and `GenerateCoreNoRewind()` to `Generator` (experimental)

### Maintenance

- Improved documentation about generator method flavors

## v2.2.0 - 2023-02-12

### Changed

- `UnmarshalBinary()` behavior so it tries to parse byte slice also as textual
  representation, not only as 128-bit byte array
- `UnmarshalText()` and `UnmarshalBinary()` of `Id` now return error instead of
  panicking when called with nil receiver

### Added

- `sql.Scanner` interface implementation to `Id`

## v2.1.2 - 2022-06-11

### Fixed

- `GenerateCore()` to update `counter_hi` when `timestamp` passed < 1000

## v2.1.1 - 2022-05-23

### Fixed

- `GenerateCore()` to reject zero as `timestamp` value

## v2.1.0 - 2022-05-22

### Added

- `GenerateCore()` and `LastStatus()` to `Generator`

### Maintenance

- Updated README

## v2.0.0 - 2022-05-01

### Changed

- Textual representation: 26-digit Base32 -> 25-digit Base36
- Field structure: { `timestamp`: 44 bits, `counter`: 28 bits, `per_sec_random`:
  24 bits, `per_gen_random`: 32 bits } -> { `timestamp`: 48 bits, `counter_hi`:
  24 bits, `counter_lo`: 24 bits, `entropy`: 32 bits }
- Timestamp epoch: 2020-01-01 00:00:00.000 UTC -> 1970-01-01 00:00:00.000 UTC
- Counter overflow handling: stall generator -> increment timestamp
- Type of generator: Generator interface -> \*Generator struct

### Removed

- `Logger` as counter overflow is no longer likely to occur
- `TimestampBias`
- `Id#Counter()`, `Id#PerSecRandom()`, `Id#PerGenRandom()`

### Added

- `Id#CounterHi()`, `Id#CounterLo()`, `Id#Entropy()`

## v1.0.0 - 2022-01-03

- Initial stable release.


id_test.go
package scru128

import (
	"bytes"
	"database/sql"
	"encoding"
	"encoding/json"
	"fmt"
	"math/big"
	"strings"
	"testing"
)

const maxUint48 uint64 = (1 << 48) - 1
const maxUint24 uint32 = (1 << 24) - 1
const maxUint32 uint32 = 0xFFFF_FFFF

// Encodes and decodes prepared cases correctly
func TestEncodeDecode(t *testing.T) {
	cases := []struct {
		timestamp uint64
		counterHi uint32
		counterLo uint32
		entropy   uint32
		string    string
	}{
		{0, 0, 0, 0, "0000000000000000000000000"},
		{maxUint48, 0, 0, 0, "F5LXX1ZZ5K6TP71GEEH2DB7K0"},
		{maxUint48, 0, 0, 0, "f5lxx1zz5k6tp71geeh2db7k0"},
		{0, maxUint24, 0, 0, "0000000005GV2R2KJWR7N8XS0"},
		{0, maxUint24, 0, 0, "0000000005gv2r2kjwr7n8xs0"},
		{0, 0, maxUint24, 0, "00000000000000JPIA7QL4HS0"},
		{0, 0, maxUint24, 0, "00000000000000jpia7ql4hs0"},
		{0, 0, 0, maxUint32, "0000000000000000001Z141Z3"},
		{0, 0, 0, maxUint32, "0000000000000000001z141z3"},
		{maxUint48, maxUint24, maxUint24, maxUint32, "F5LXX1ZZ5PNORYNQGLHZMSP33"},
		{maxUint48, maxUint24, maxUint24, maxUint32, "f5lxx1zz5pnorynqglhzmsp33"},
	}

	for _, e := range cases {
		var fromFields, fromString Id
		fromFields = FromFields(
			e.timestamp, e.counterHi, e.counterLo, e.entropy,
		)
		fromString, _ = Parse(e.string)

		caseStringAsBigInt, _ := new(big.Int).SetString(e.string, 36)
		if new(big.Int).SetBytes(fromFields[:]).Cmp(caseStringAsBigInt) != 0 ||
			fromFields.Timestamp() != e.timestamp ||
			fromFields.CounterHi() != e.counterHi ||
			fromFields.CounterLo() != e.counterLo ||
			fromFields.Entropy() != e.entropy ||
			fromFields.String() != strings.ToLower(e.string) {
			t.Fail()
		}
		if new(big.Int).SetBytes(fromString[:]).Cmp(caseStringAsBigInt) != 0 ||
			fromString.Timestamp() != e.timestamp ||
			fromString.CounterHi() != e.counterHi ||
			fromString.CounterLo() != e.counterLo ||
			fromString.Entropy() != e.entropy ||
			fromString.String() != strings.ToLower(e.string) {
			t.Fail()
		}
	}
}

// Returns error if an invalid string representation is supplied
func TestStringValidation(t *testing.T) {
	cases := []string{
		"",
		" 036z8puq4tsxsigk6o19y164q",
		"036z8puq54qny1vq3hcbrkweb ",
		" 036z8puq54qny1vq3helivwax ",
		"+036z8puq54qny1vq3hfcv3ss0",
		"-036z8puq54qny1vq3hhy8u1ch",
		"+36z8puq54qny1vq3hjq48d9p",
		"-36z8puq5a7j0ti08oz6zdrdy",
		"036z8puq5a7j0t_08p2cdz28v",
		"036z8pu-5a7j0ti08p3ol8ool",
		"036z8puq5a7j0ti08p4j 6cya",
		"f5lxx1zz5pnorynqglhzmsp34",
		"zzzzzzzzzzzzzzzzzzzzzzzzz",
		"039o\tvvklfmqlqe7fzllz7c7t",
		"039onvvklfmqlq漢字fgvd1",
		"039onvvkl🤣qe7fzr2hdoqu",
		"頭onvvklfmqlqe7fzrhtgcfz",
		"039onvvklfmqlqe7fztft5尾",
		"039漢字a52xp4bvf4sn94e09cja",
		"039ooa52xp4bv😘sn97642mwl",
	}

	for _, e := range cases {
		var err error
		_, err = Parse(e)
		if err == nil {
			t.Fail()
		}
	}
}

// Has symmetric converters from/to various values
func TestSymmetricConverters(t *testing.T) {
	cases := []Id{
		FromFields(0, 0, 0, 0),
		FromFields(maxUint48, 0, 0, 0),
		FromFields(0, maxUint24, 0, 0),
		FromFields(0, 0, maxUint24, 0),
		FromFields(0, 0, 0, maxUint32),
		FromFields(maxUint48, maxUint24, maxUint24, maxUint32),
	}

	g := NewGenerator()
	for i := 0; i < 1_000; i++ {
		e, _ := g.Generate()
		cases = append(cases, e)
	}

	for _, e := range cases {
		if x, _ := Parse(e.String()); x != e {
			t.Fail()
		}
		if FromFields(
			e.Timestamp(), e.CounterHi(), e.CounterLo(), e.Entropy(),
		) != e {
			t.Fail()
		}

		marshaledBinary, _ := e.MarshalBinary()
		marshaledText, _ := e.MarshalText()
		unmarshaled := new(Id)
		if unmarshaled.UnmarshalBinary(marshaledBinary) != nil || *unmarshaled != e {
			t.Fail()
		}
		if unmarshaled.UnmarshalBinary(marshaledText) != nil || *unmarshaled != e {
			t.Fail()
		}
		if unmarshaled.UnmarshalText(marshaledText) != nil || *unmarshaled != e {
			t.Fail()
		}

		scanned := new(Id)
		if scanned.Scan(e.String()) != nil || *scanned != e {
			t.Fail()
		}
		if scanned.Scan(marshaledBinary) != nil || *scanned != e {
			t.Fail()
		}
		if scanned.Scan(marshaledText) != nil || *scanned != e {
			t.Fail()
		}
	}
}

// Supports comparison methods
func TestComparisonMethods(t *testing.T) {
	ordered := []Id{
		FromFields(0, 0, 0, 0),
		FromFields(0, 0, 0, 1),
		FromFields(0, 0, 0, maxUint32),
		FromFields(0, 0, 1, 0),
		FromFields(0, 0, maxUint24, 0),
		FromFields(0, 1, 0, 0),
		FromFields(0, maxUint24, 0, 0),
		FromFields(1, 0, 0, 0),
		FromFields(2, 0, 0, 0),
	}

	g := NewGenerator()
	for i := 0; i < 1_000; i++ {
		e, _ := g.Generate()
		ordered = append(ordered, e)
	}

	prev := ordered[0]
	for _, curr := range ordered[1:] {
		if curr == prev || curr.Cmp(prev) < 0 || prev.Cmp(curr) > 0 {
			t.Fail()
		}

		clone := curr
		if curr != clone || curr.Cmp(clone) != 0 || clone.Cmp(curr) != 0 {
			t.Fail()
		}

		prev = curr
	}
}

// Serializes and deserializes an object using the canonical string
// representation
func TestSerializedForm(t *testing.T) {
	g := NewGenerator()
	for i := 0; i < 1_000; i++ {
		obj, _ := g.Generate()
		strJson := []byte(`"` + obj.String() + `"`)

		marshaled, _ := json.Marshal(obj)
		unmarshaled := new(Id)
		json.Unmarshal(strJson, unmarshaled)
		if !bytes.Equal(marshaled, strJson) || *unmarshaled != obj {
			t.Fail()
		}
	}
}

// Ensures compliance with interfaces.
func TestInterfaces(t *testing.T) {
	var x Id
	var _ fmt.Stringer = x
	var _ encoding.TextMarshaler = x
	var _ encoding.TextUnmarshaler = &x
	var _ encoding.BinaryMarshaler = x
	var _ encoding.BinaryUnmarshaler = &x
	var _ sql.Scanner = &x
}


gen_test.go
package scru128

import (
	"bufio"
	crand "crypto/rand"
	mrand "math/rand"
	"testing"
	"time"
)

// Generates increasing IDs even with decreasing or constant timestamp
func TestDecreasingOrConstantTimestampReset(t *testing.T) {
	var ts uint64 = 0x0123_4567_89ab
	var g *Generator = NewGenerator()

	prev, _ := g.GenerateOrResetCore(ts, 10_000)
	if prev.Timestamp() != ts {
		t.Fail()
	}

	for i := uint64(0); i < 100_000; i++ {
		var curr Id
		if i < 9_999 {
			curr, _ = g.GenerateOrResetCore(ts-i, 10_000)
		} else {
			curr, _ = g.GenerateOrResetCore(ts-9_999, 10_000)
		}
		if prev.Cmp(curr) >= 0 {
			t.Fail()
		}
		prev = curr
	}
	if prev.Timestamp() < ts {
		t.Fail()
	}
}

// Breaks increasing order of IDs if timestamp went backwards a lot
func TestTimestampRollbackReset(t *testing.T) {
	var ts uint64 = 0x0123_4567_89ab
	var g *Generator = NewGenerator()

	prev, _ := g.GenerateOrResetCore(ts, 10_000)
	if prev.Timestamp() != ts {
		t.Fail()
	}

	curr, _ := g.GenerateOrResetCore(ts-10_000, 10_000)
	if prev.Cmp(curr) >= 0 {
		t.Fail()
	}

	prev = curr
	curr, _ = g.GenerateOrResetCore(ts-10_001, 10_000)
	if prev.Cmp(curr) <= 0 {
		t.Fail()
	}
	if curr.Timestamp() != ts-10_001 {
		t.Fail()
	}

	prev = curr
	curr, _ = g.GenerateOrResetCore(ts-10_002, 10_000)
	if prev.Cmp(curr) >= 0 {
		t.Fail()
	}
}

// Generates increasing IDs even with decreasing or constant timestamp
func TestDecreasingOrConstantTimestampAbort(t *testing.T) {
	var ts uint64 = 0x0123_4567_89ab
	var g *Generator = NewGenerator()

	prev, err := g.GenerateOrAbortCore(ts, 10_000)
	if err == ErrClockRollback {
		t.Fail()
	}
	if prev.Timestamp() != ts {
		t.Fail()
	}

	for i := uint64(0); i < 100_000; i++ {
		var curr Id
		if i < 9_999 {
			curr, err = g.GenerateOrAbortCore(ts-i, 10_000)
		} else {
			curr, err = g.GenerateOrAbortCore(ts-9_999, 10_000)
		}
		if err == ErrClockRollback {
			t.Fail()
		}
		if prev.Cmp(curr) >= 0 {
			t.Fail()
		}
		prev = curr
	}
	if prev.Timestamp() < ts {
		t.Fail()
	}
}

// Returns error if timestamp went backwards a lot
func TestTimestampRollbackAbort(t *testing.T) {
	var ts uint64 = 0x0123_4567_89ab
	var g *Generator = NewGenerator()

	prev, err := g.GenerateOrAbortCore(ts, 10_000)
	if err == ErrClockRollback {
		t.Fail()
	}
	if prev.Timestamp() != ts {
		t.Fail()
	}

	curr, err := g.GenerateOrAbortCore(ts-10_000, 10_000)
	if err == ErrClockRollback {
		t.Fail()
	}
	if prev.Cmp(curr) >= 0 {
		t.Fail()
	}

	_, err = g.GenerateOrAbortCore(ts-10_001, 10_000)
	if err != ErrClockRollback {
		t.Fail()
	}

	_, err = g.GenerateOrAbortCore(ts-10_002, 10_000)
	if err != ErrClockRollback {
		t.Fail()
	}
}

func BenchmarkGeneratorDefault(b *testing.B) {
	g := NewGenerator()
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		g.Generate()
	}
}

func BenchmarkGeneratorBufferedCryptoRand(b *testing.B) {
	g := NewGeneratorWithRng(bufio.NewReader(crand.Reader))
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		g.Generate()
	}
}

func BenchmarkGeneratorInsecureMathRand(b *testing.B) {
	g := NewGeneratorWithRng(mrand.New(mrand.NewSource(time.Now().UnixMilli())))
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		g.Generate()
	}
}


scru128_test.go
package scru128

import (
	"fmt"
	"math"
	"regexp"
	"sync"
	"testing"
	"time"
)

var samples = make([]string, 100_000)

func init() {
	for i := range samples {
		samples[i] = NewString()
	}
}

// Generates 25-digit canonical string
func TestFormat(t *testing.T) {
	re := regexp.MustCompile(`^[0-9a-z]{25}$`)
	for _, e := range samples {
		if !re.MatchString(e) {
			t.Fail()
		}
	}
}

// Generates 100k identifiers without collision
func TestUniqueness(t *testing.T) {
	set := make(map[string]struct{}, len(samples))
	for _, e := range samples {
		set[e] = struct{}{}
	}
	if len(set) != len(samples) {
		t.Fail()
	}
}

// Generates sortable string representation by creation time
func TestOrder(t *testing.T) {
	for i := 1; i < len(samples); i++ {
		if samples[i-1] >= samples[i] {
			t.Fail()
		}
	}
}

// Encodes up-to-date timestamp
func TestTimestamp(t *testing.T) {
	var g *Generator = NewGenerator()
	for i := 0; i < 10_000; i++ {
		tsNow := time.Now().UnixMilli()
		x, _ := g.Generate()
		if math.Abs(float64(tsNow-int64(x.Timestamp()))) >= 16 {
			t.Fail()
		}
	}
}

// Encodes unique sortable tuple of timestamp and counters
func TestTimestampAndCounters(t *testing.T) {
	prev, _ := Parse(samples[0])
	for _, e := range samples[1:] {
		curr, _ := Parse(e)
		if !(prev.Timestamp() < curr.Timestamp() ||
			(prev.Timestamp() == curr.Timestamp() &&
				prev.CounterHi() < curr.CounterHi()) ||
			(prev.Timestamp() == curr.Timestamp() &&
				prev.CounterHi() == curr.CounterHi() &&
				prev.CounterLo() < curr.CounterLo())) {
			t.Fail()
		}
		prev = curr
	}
}

// Generates no IDs sharing same timestamp and counters under multithreading
func TestThreading(t *testing.T) {
	results := make(chan Id, 4*10_000)

	group := new(sync.WaitGroup)
	for i := 0; i < 4; i++ {
		group.Add(1)
		go func() {
			defer group.Done()
			for i := 0; i < 10_000; i++ {
				results <- New()
			}
		}()
	}

	done := make(chan struct{})
	go func() {
		defer close(done)
		set := make(map[string]struct{}, 4*10_000)
		for e := range results {
			set[fmt.Sprintf("%012x-%06x-%06x", e.Timestamp(), e.CounterHi(), e.CounterLo())] = struct{}{}
		}
		if len(set) != 4*10_000 {
			t.Fail()
		}
	}()

	group.Wait()
	close(results)
	<-done
}

func BenchmarkNewString(b *testing.B) {
	b.ResetTimer()
	for i := 0; i < b.N; i++ {
		NewString()
	}
}


scru128.go
// SCRU128: Sortable, Clock and Random number-based Unique identifier
//
// SCRU128 ID is yet another attempt to supersede UUID for the users who need
// decentralized, globally unique time-ordered identifiers. SCRU128 is inspired
// by ULID and KSUID and has the following features:
//
//   - 128-bit unsigned integer type
//   - Sortable by generation time (as integer and as text)
//   - 25-digit case-insensitive textual representation (Base36)
//   - 48-bit millisecond Unix timestamp that ensures useful life until year
//     10889
//   - Up to 281 trillion time-ordered but unpredictable unique IDs per
//     millisecond
//   - 80-bit three-layer randomness for global uniqueness
//
// See SCRU128 Specification for details: https://github.com/scru128/spec
package scru128

// The maximum value of 48-bit timestamp field.
const maxTimestamp uint64 = 0xffff_ffff_ffff

// The maximum value of 24-bit counter_hi field.
const maxCounterHi uint32 = 0xff_ffff

// The maximum value of 24-bit counter_lo field.
const maxCounterLo uint32 = 0xff_ffff

var globalGenerator = NewGenerator()

// Generates a new SCRU128 ID object using the global generator, or panics if
// crypto/rand fails.
//
// This function is thread-safe; multiple threads can call it concurrently.
func New() Id {
	id, err := globalGenerator.Generate()
	if err != nil {
		panic(err)
	}
	return id
}

// Generates a new SCRU128 ID encoded in the 26-digit canonical string
// representation using the global generator, or panics if crypto/rand fails.
//
// This function is thread-safe. Use this to quickly get a new SCRU128 ID as a
// string.
func NewString() string {
	return New().String()
}


generator.go
package scru128

import (
	"bufio"
	"crypto/rand"
	"fmt"
	"io"
	"sync"
	"time"
)

// Represents a SCRU128 ID generator that encapsulates the monotonic counters
// and other internal states.
//
// This structure must be instantiated by one of the dedicated constructors:
// [NewGenerator] or [NewGeneratorWithRng].
//
// # Generator functions
//
// The generator comes with four different methods that generate a SCRU128 ID:
//
//	| Flavor              | Timestamp | Thread- | On big clock rewind |
//	| ------------------- | --------- | ------- | ------------------- |
//	| Generate            | Now       | Safe    | Resets generator    |
//	| GenerateOrAbort     | Now       | Safe    | Returns error       |
//	| GenerateOrResetCore | Argument  | Unsafe  | Resets generator    |
//	| GenerateOrAbortCore | Argument  | Unsafe  | Returns error       |
//
// All of the four return a monotonically increasing ID by reusing the previous
// `timestamp` even if the one provided is smaller than the immediately
// preceding ID's. However, when such a clock rollback is considered significant
// (by default, more than ten seconds):
//
//  1. `Generate` (OrReset) methods reset the generator and return a new ID
//     based on the given `timestamp`, breaking the increasing order of IDs.
//  2. `OrAbort` variants abort and return the [ErrClockRollback] error value
//     immediately.
//
// The `Core` functions offer low-level thread-unsafe primitives to customize
// the behavior.
type Generator struct {
	timestamp uint64
	counterHi uint32
	counterLo uint32

	// The timestamp at the last renewal of counter_hi field.
	tsCounterHi uint64

	// The random number generator used by the generator.
	rng io.Reader

	lock sync.Mutex
}

// Creates a generator object with the default random number generator.
//
// The crypto/rand random number generator is quite slow for small reads on some
// platforms. In such a case, wrapping crypto/rand with bufio.Reader may result
// in a drastic improvement in the throughput of generator. If the throughput is
// an important issue, check out the following benchmark tests and pass
// bufio.NewReader(rand.Reader) to [NewGeneratorWithRng]:
//
//	go test -bench Generator
func NewGenerator() *Generator {
	// use small buffer by default to avoid both occasional unbearable performance
	// degradation and waste of time and space for unused buffer contents
	br := bufio.NewReaderSize(rand.Reader, 32)
	return NewGeneratorWithRng(br)
}

// Creates a generator object with a specified random number generator. The
// specified random number generator should be cryptographically strong and
// securely seeded.
//
// This constructor panics if `rng` is nil.
func NewGeneratorWithRng(rng io.Reader) *Generator {
	if rng == nil {
		panic("constructor called with nil `rng`")
	}
	return &Generator{rng: rng}
}

// Generates a new SCRU128 ID object from the current `timestamp`, or resets the
// generator upon significant timestamp rollback.
//
// See the [Generator] type documentation for the description.
//
// This method returns a non-nil err if the random number generator fails.
func (g *Generator) Generate() (id Id, err error) {
	g.lock.Lock()
	defer g.lock.Unlock()
	return g.GenerateOrResetCore(
		uint64(time.Now().UnixMilli()),
		defaultRollbackAllowance,
	)
}

// Generates a new SCRU128 ID object from the current `timestamp`, or returns an
// error upon significant timestamp rollback.
//
// See the [Generator] type documentation for the description.
//
// This method returns a non-nil err if the random number generator fails or
// returns the [ErrClockRollback] err upon significant clock rollback.
func (g *Generator) GenerateOrAbort() (id Id, err error) {
	g.lock.Lock()
	defer g.lock.Unlock()
	return g.GenerateOrAbortCore(
		uint64(time.Now().UnixMilli()),
		defaultRollbackAllowance,
	)
}

// Generates a new SCRU128 ID object from the `timestamp` passed, or resets the
// generator upon significant timestamp rollback.
//
// See the [Generator] type documentation for the description.
//
// The `rollbackAllowance` parameter specifies the amount of timestamp rollback
// that is considered significant. A suggested value is `10_000` (milliseconds).
//
// Unlike [Generator.Generate], this method is NOT thread-safe. The generator
// object should be protected from concurrent accesses using a mutex or other
// synchronization mechanism to avoid race conditions.
//
// This method returns a non-nil err if the random number generator fails.
//
// This method panics if `timestamp` is not a 48-bit positive integer.
func (g *Generator) GenerateOrResetCore(
	timestamp uint64,
	rollbackAllowance uint64,
) (id Id, err error) {
	id, err = g.GenerateOrAbortCore(timestamp, rollbackAllowance)
	if err == ErrClockRollback {
		// reset state and resume
		g.timestamp = 0
		g.tsCounterHi = 0
		id, err = g.GenerateOrAbortCore(timestamp, rollbackAllowance)
	}
	return
}

// Generates a new SCRU128 ID object from the `timestamp` passed, or returns an
// error upon significant timestamp rollback.
//
// See the [Generator] type documentation for the description.
//
// The `rollbackAllowance` parameter specifies the amount of timestamp rollback
// that is considered significant. A suggested value is `10_000` (milliseconds).
//
// Unlike [Generator.GenerateOrAbort], this method is NOT thread-safe. The
// generator object should be protected from concurrent accesses using a mutex
// or other synchronization mechanism to avoid race conditions.
//
// This method returns a non-nil err if the random number generator fails or
// returns the [ErrClockRollback] err upon significant clock rollback.
//
// This method panics if `timestamp` is not a 48-bit positive integer.
func (g *Generator) GenerateOrAbortCore(
	timestamp uint64,
	rollbackAllowance uint64,
) (id Id, err error) {
	if g == nil || g.rng == nil {
		panic("method call on invalid receiver")
	} else if timestamp == 0 || timestamp > maxTimestamp {
		panic("`timestamp` must be a 48-bit positive integer")
	} else if rollbackAllowance > maxTimestamp {
		panic("`rollbackAllowance` out of reasonable range")
	}

	var n uint32
	if timestamp > g.timestamp {
		g.timestamp = timestamp
		n, err = g.randomUint32()
		if err != nil {
			return Id{}, err
		}
		g.counterLo = n & maxCounterLo
	} else if timestamp+rollbackAllowance >= g.timestamp {
		// go on with previous timestamp if new one is not much smaller
		g.counterLo++
		if g.counterLo > maxCounterLo {
			g.counterLo = 0
			g.counterHi++
			if g.counterHi > maxCounterHi {
				g.counterHi = 0
				// increment timestamp at counter overflow
				g.timestamp++
				n, err = g.randomUint32()
				if err != nil {
					return Id{}, err
				}
				g.counterLo = n & maxCounterLo
			}
		}
	} else {
		// abort if clock went backwards to unbearable extent
		return Id{}, ErrClockRollback
	}

	if g.timestamp-g.tsCounterHi >= 1_000 || g.tsCounterHi == 0 {
		g.tsCounterHi = g.timestamp
		n, err = g.randomUint32()
		if err != nil {
			return Id{}, err
		}
		g.counterHi = n & maxCounterHi
	}

	n, err = g.randomUint32()
	if err != nil {
		return Id{}, err
	}
	return FromFields(g.timestamp, g.counterHi, g.counterLo, n), nil
}

// The default timestamp rollback allowance.
const defaultRollbackAllowance = 10_000 // 10 seconds

// The error value returned by [Generator.GenerateOrAbort] and
// [Generator.GenerateOrAbortCore] when the relevant timestamp is significantly
// smaller than the one embedded in the immediately preceding ID generated by
// the generator.
var ErrClockRollback = fmt.Errorf(
	"scru128.Generator: detected unbearable clock rollback")

// Returns a random uint32 value.
func (g *Generator) randomUint32() (uint32, error) {
	b := make([]byte, 4)
	_, err := g.rng.Read(b)
	if err != nil {
		err = fmt.Errorf("scru128.Generator: random number generator error: %w", err)
	}
	_ = b[3] // bounds check hint to compiler
	return uint32(b[3]) | uint32(b[2])<<8 | uint32(b[1])<<16 | uint32(b[0])<<24, err
}
